src/display/annotation_storage.js

```diff
@@ -148,6 +151,20 @@ class AnnotationStorage {
     return clone;
   }
 
+  /**
+   * PLEASE NOTE: Only intended for usage within the API itself.
+   * @ignore
+   */
+  get frozen() {
+    if (this.#frozen) {
+      const results = this.#frozen;
+      this.#frozen = null;
+      return results;
+    }
```

Comment:
How does this work when printing a *multi-page* document, since it seems to me that the very first `getOperatorlist`-call will remove the internal frozen-state?

Generally speaking, I'm sorry to say that I'm also finding this implementation a bit cumbersome and unfortunate.
If you don't mind, I'd really like to experiment a little bit and maybe propose a slightly different solution here (hopefully later today or tomorrow).

Reply:
Please note that I'm not saying that this is perfect by any means, and it feels a bit annoying having to extend the public API, but what about something like the following?
https://github.com/mozilla/pdf.js/compare/master...Snuffleupagus:PrintAnnotationStorage?w=1